addons/binding/org.openhab.binding.km200/src/main/java/org/openhab/binding/km200/handler/KM200GatewayHandler.java

```diff
@@ -0,0 +1,710 @@
+            String rootPath = tType.getRootPath();
+            if (rootPath.isEmpty()) {
+                continue;
+            }
+            logger.debug("Add Property rootPath: {}", rootPath);
+            KM200CommObject serObj = getDevice().getServiceObject(rootPath);
+            for (String subKey : asProperties) {
+                if (serObj.serviceTreeMap.containsKey(subKey)) {
+                    String subKeyType = serObj.serviceTreeMap.get(subKey).getServiceType();
+                    if (subKeyType.compareTo("stringValue") != 0 && subKeyType.compareTo("floatValue") != 0) {
+                        continue;
+                    }
+                    if (bridgeProperties.containsKey(subKey)) {
+                        bridgeProperties.remove(subKey);
+                    }
+                    logger.debug("Add Property: {}  :{}", subKey,
+                            serObj.serviceTreeMap.get(subKey).getValue().toString());
+                    bridgeProperties.put(subKey, serObj.serviceTreeMap.get(subKey).getValue().toString());
+                }
+            }
+        }
+        updateProperties(bridgeProperties);
+    }
+
+    /**
+     * This function checks whether the service has a replacement parameter
+     *
+     * @param channel
+     * @param device
+     */
+    public static String checkParameterReplacement(Channel channel, KM200Device device) {
+        String service = translatesNameToPath(channel.getProperties().get("root"));
+        if (service.contains(SWITCH_PROGRAM_REPLACEMENT)) {
+            String currentService = translatesNameToPath(channel.getProperties().get(SWITCH_PROGRAM_CURRENT_PATH_NAME));
+            if (device.containsService(currentService)) {
+                if (device.getServiceObject(currentService).getServiceType().equals("stringValue")) {
+                    String val = (String) device.getServiceObject(currentService).getValue();
+                    service = service.replace("__current__", val);
+                    return service;
+                }
+            }
+        }
+        return service;
+    }
+
+    /**
+     * This function checks whether the channel has channel parameters
+     *
+     * @param channel
+     */
+    @SuppressWarnings("null")
+    public static Map<String, String> getChannelConfigurationStrings(Channel channel) {
+
+        Map<String, String> paraNames = new HashMap<String, String>();
+        if (channel.getConfiguration() != null) {
+            if (channel.getConfiguration().containsKey("on")) {
+                paraNames.put("on", channel.getConfiguration().get("on").toString());
+                logger.debug("Added ON: {}", channel.getConfiguration().get("on").toString());
+            }
+
+            if (channel.getConfiguration().containsKey("off")) {
+                paraNames.put("off", channel.getConfiguration().get("off").toString());
+                logger.debug("Added OFF: {}", channel.getConfiguration().get("off").toString());
```

Comment:
You don't have to call `toString` it will be called automatically, but only when the logging-level actually applies.

Please check all